### DISCUSSION TITLE
Nullability test in C# 9

### DISCUSSION THREAD

**Discussion by sjb-sjb** [dotnet/csharplang#4176] [General]
Just taking C# 9 for a spin and found this -- 

        protected void RaiseCollectionChanged(NotifyCollectionChangedEventArgs eventArgs, DeferralSource? deferralSource = null)
        {
            Debug.Assert(eventArgs.NewItems == null || eventArgs.NewItems.Count == 0 || eventArgs.NewItems[0] == null || eventArgs.NewItems[0].GetType() == typeof(TEntity));
            ...
        }

Expected: no warnings
Actual: "Dereference of a possibly null reference" in the expression "eventArgs.NewItems[0].GetType()". 

It seems to think that eventArgs.NewItems[0] could be null even though that possibility was just tested in the preceding OR clause. Weirdly, this happens on eventArgs.NewItems[0].GetType() but not on eventArgs.NewItems.Count.

**Comment by YairHalberstadt** [comment]
Can the element type of New items be null?

**Comment by jnm2** [comment]
I think this is because nullability is remembered for simple property access, treating properties like variables, it's not remembered for indexers. Similarly, it's not remembered for method calls. Here's how I would prefer to write the assertion:

```cs
Debug.Assert(!(eventArgs.NewItems is { Count: > 0 } && eventArgs.NewItems[0] is { } first && first.GetType() != typeof(TEntity)));

// Or
Debug.Assert(!(eventArgs.NewItems?.Cast<object>().FirstOrDefault() is { } first && first.GetType() != typeof(TEntity)));
```

If the comparison should be `eventArgs.NewItems[0] is TEntity` rather than `eventArgs.NewItems[0].GetType() == typeof(TEntity)` (the difference matters whenever `TEntity` is not known to be sealed) then this becomes possible:

```cs
Debug.Assert(eventArgs.NewItems?.Cast<object>().FirstOrDefault() is null or TEntity);
```
